receiver/azureblobrehydrationreceiver/receiver.go

```diff
@@ -0,0 +1,417 @@
+	for ; i < len(parts)-1; i++ {
+		part := parts[i]
+
+		if !strings.HasPrefix(part, nextExpectedPart) {
+			err = errInvalidBlobPath
+			return
+		}
+
+		val := strings.TrimPrefix(part, nextExpectedPart)
+
+		switch nextExpectedPart {
+		case year:
+			nextExpectedPart = month
+		case month:
+			tsBuilder.WriteString("-")
+			nextExpectedPart = day
+		case day:
+			tsBuilder.WriteString("-")
+			nextExpectedPart = hour
+		case hour:
+			tsBuilder.WriteString("T")
+			nextExpectedPart = minute
+		case minute:
+			tsBuilder.WriteString(":")
+			nextExpectedPart = ""
+		}
+
+		tsBuilder.WriteString(val)
+	}
+
+	// Special case when using hour granularity.
+	// There won't be a minute=XX part of the path if we've exited the loop
+	// and we still expect minutes just write ':00' for minutes.
+	if nextExpectedPart == minute {
+		tsBuilder.WriteString(":00")
+	}
+
+	// Parse the expected format
+	parsedTime, timeErr := time.Parse(timeFormat, tsBuilder.String())
+	if timeErr != nil {
+		err = fmt.Errorf("parse blob time: %w", timeErr)
+		return
+	}
+	blobTime = &parsedTime
+
+	// For the last part of the path parse the telemetry type
+	lastPart := parts[len(parts)-1]
+	switch {
+	case strings.Contains(lastPart, metricBlobSignifier):
+		telemetryType = component.DataTypeMetrics
+	case strings.Contains(lastPart, logsBlobSignifier):
+		telemetryType = component.DataTypeLogs
+	case strings.Contains(lastPart, tracesBlobSignifier):
+		telemetryType = component.DataTypeTraces
+	}
+
+	return
+}
+
+// isInTimeRange returns true if startingTime <= blobTime <= endingTime
+func (r *rehydrationReceiver) isInTimeRange(blobTime time.Time) bool {
+	return (blobTime.Equal(r.startingTime) || blobTime.After(r.startingTime)) &&
+		(blobTime.Equal(r.endingTime) || blobTime.Before(r.endingTime))
+}
+
+// processBlob does the following:
+// 1. Downloads the blob
+// 2. Decompresses the blob if applicable
+// 3. Pass the blob to the consumer
+func (r *rehydrationReceiver) processBlob(blob *azureblob.BlobInfo) error {
+	// Allocate a buffer the size of the blob. If the buffer isn't big enough download errors.
+	blobBuffer := make([]byte, blob.Size)
+
+	size, err := r.azureClient.DownloadBlob(r.ctx, r.cfg.Container, blob.Name, blobBuffer)
+	if err != nil {
+		return fmt.Errorf("download blob: %w", err)
+	}
+
+	// Check file extension see if we need to decompress
+	ext := filepath.Ext(blob.Name)
+	switch {
+	case ext == ".gz":
+		blobBuffer, err = gzipDecompress(blobBuffer[:size])
+		if err != nil {
+			return fmt.Errorf("gzip: %w", err)
+		}
+	case ext == ".json":
+		// Do nothing for json files
+	default:
+		return fmt.Errorf("unsupported file type: %s", ext)
+	}
+
+	if err := r.consumer.Consume(r.ctx, blobBuffer); err != nil {
+		return fmt.Errorf("consume: %w", err)
+	}
+	return nil
+}
+
+// gzipDecompress does a gzip decompression on the passed in contents
+func gzipDecompress(contents []byte) ([]byte, error) {
+	gr, err := gzip.NewReader(bytes.NewBuffer(contents))
```

Comment:
We should close this and check the error at the end of the function, I think it's possible close returns a checksum error.